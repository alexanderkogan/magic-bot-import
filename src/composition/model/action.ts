export interface Action<Name = string> {
    type: Name
}
