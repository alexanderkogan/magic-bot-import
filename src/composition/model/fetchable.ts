export type Fetchable<T> = Empty | Loading | Loaded<T> | Failed

export interface Empty {
    type: 'empty'
}
export interface Loading {
    type: 'loading'
}
export interface Loaded<T> {
    type: 'loaded'
    value: T
}
export interface Failed {
    type: 'failed'
    value: Error
}

export function empty(): Empty {
    return { type: 'empty' }
}
export function loading(): Loading {
    return { type: 'loading' }
}
export function loaded<T>(value: T): Loaded<T> {
    return { type: 'loaded', value }
}
export function failed(value: Error): Failed {
    return { type: 'failed', value }
}
