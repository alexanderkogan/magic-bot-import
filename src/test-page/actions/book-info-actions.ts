import { Action } from '../../composition/model/action'

export type BookInfoAction = BookInfoFetchedAction | BookInfoFetchFailedAction | BookInfoFetchingAction

export type BookInfoFetchingAction = Action<'BOOK_INFO_FETCHING'>
export type BookInfoFetchedAction = Action<'BOOK_INFO_FETCHED'> & { value: string }
export type BookInfoFetchFailedAction = Action<'BOOK_INFO_FETCH_FAILED'> & { error: Error }

export function fetching(): BookInfoFetchingAction {
    return { type: 'BOOK_INFO_FETCHING' }
}
export function fetched(value: string): BookInfoFetchedAction {
    return { type: 'BOOK_INFO_FETCHED', value }
}
export function failed(error: Error): BookInfoFetchFailedAction {
    return { type: 'BOOK_INFO_FETCH_FAILED', error }
}
