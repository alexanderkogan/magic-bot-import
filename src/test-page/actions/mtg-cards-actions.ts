import { Action } from '../../composition/model/action'
import { Card } from 'scryfall-sdk'

export type MtgCardsAction = MtgCardsFetchedAction | MtgCardsFetchFailedAction | MtgCardsFetchingAction

export type MtgCardsFetchingAction = Action<'MTG_CARDS_FETCHING'>
export type MtgCardsFetchedAction = Action<'MTG_CARDS_FETCHED'> & { value: Card[] }
export type MtgCardsFetchFailedAction = Action<'MTG_CARDS_FETCH_FAILED'> & { error: Error }

export function fetching(): MtgCardsFetchingAction {
    return { type: 'MTG_CARDS_FETCHING' }
}
export function fetched(value: Card[]): MtgCardsFetchedAction {
    return { type: 'MTG_CARDS_FETCHED', value }
}
export function failed(error: Error): MtgCardsFetchFailedAction {
    return { type: 'MTG_CARDS_FETCH_FAILED', error }
}
