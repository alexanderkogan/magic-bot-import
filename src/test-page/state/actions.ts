import { Action } from '../../composition/model/action'
import { SelectableValue } from '.'

export type TestItemAction = PopTestItemAction | PushTestItemAction

export type PopTestItemAction = Action<'TEST_ITEM_POP'>
export type PushTestItemAction = Action<'TEST_ITEM_PUSH'> & { value: SelectableValue }

export function pushItem(value: SelectableValue): PushTestItemAction {
    return { type: 'TEST_ITEM_PUSH', value }
}

export function popItem(): PopTestItemAction {
    return { type: 'TEST_ITEM_POP' }
}
