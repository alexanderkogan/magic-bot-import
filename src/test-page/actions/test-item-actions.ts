import { Action } from '../../composition/model/action'
import { SelectableValue } from '../state'

export type TestItemAction = PopTestItemAction | PushTestItemAction

export type PopTestItemAction = Action<'TEST_ITEM_POP'>
export type PushTestItemAction = Action<'TEST_ITEM_PUSH'> & { value: SelectableValue }

export function push(value: SelectableValue): PushTestItemAction {
    return { type: 'TEST_ITEM_PUSH', value }
}

export function pop(): PopTestItemAction {
    return { type: 'TEST_ITEM_POP' }
}
