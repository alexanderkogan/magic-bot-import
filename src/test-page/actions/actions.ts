import { TestItemAction } from './test-item-actions'
import { BookInfoAction } from './book-info-actions'

export type TestPageAction = TestItemAction | BookInfoAction
