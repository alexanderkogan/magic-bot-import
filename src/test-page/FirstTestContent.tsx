import * as React from 'react'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { Props } from '../react-typings'

import { popItem, pushItem } from './state/actions'

import Content from '../pattern-library/Content'
import Button from '../pattern-library/Button'
import Icon from '../pattern-library/Icon'
import Divider from '../pattern-library/Divider'
import ToggleButton from '../pattern-library/ToggleButton'
import { RootState, extractTestPage } from '../composition/store'
import { TestPageState, SelectableValue } from './state'

export interface FirstTestContentProps {
    push: (name: SelectableValue) => {}
    pop: () => {}
    selections: number
    lastSelection: SelectableValue | undefined
}

export const mapStateToProps = (state: TestPageState) => ({
    selections: state.valuesSelected.length,
    lastSelection: state.valuesSelected[state.valuesSelected.length - 1],
})

export const dispatchToProps = (dispatch: Dispatch) => ({
    push: (name: SelectableValue) => dispatch(pushItem(name)),
    pop: () => dispatch(popItem()),
})

export const component = (props: Props<FirstTestContentProps>) => {
    return (
        <Content>
            <Button type="primary" clicked={props.pop}>
                delete last item <Icon type="trash" />
            </Button>
            <Divider />
            last selection: {props.lastSelection || 'nothing selected yet'}
            <ToggleButton
                id="valueToggle"
                values={['first value', 'second value', 'another value']}
                changed={props.push}
                currentValue={props.lastSelection}
            ></ToggleButton>
        </Content>
    )
}

export default connect(
    (state: RootState) => mapStateToProps(extractTestPage(state)),
    dispatch => dispatchToProps(dispatch)
)(component)
