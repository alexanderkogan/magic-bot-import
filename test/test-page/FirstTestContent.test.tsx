import * as React from 'react'

import { component as FirstTestContent, mapStateToProps, dispatchToProps } from '../../src/test-page/FirstTestContent'
import { shallow } from '../test-utils'
import * as TestItemActions from '../../src/test-page/actions/test-item-actions'
import { empty } from '../../src/composition/model/fetchable'

describe('FirstTestContent', () => {
    describe('state mapping', () => {
        it('maps empty state to props', () => {
            const props = mapStateToProps({ valuesSelected: [], bookInfo: empty() })
            expect(props).toEqual({ selections: 0 })
        })
        it('maps filled state to props', () => {
            const props = mapStateToProps({ valuesSelected: ['first value', 'another value'], bookInfo: empty() })
            expect(props).toEqual({ selections: 2, lastSelection: 'another value' })
        })
    })

    describe('dispatch mapping', () => {
        it('dispatches pop action', async () => {
            const dispatcher = jest.fn()
            dispatchToProps(dispatcher).pop()
            expect(dispatcher).toHaveBeenCalledWith(TestItemActions.pop())
        })

        it('dispatches push action', async () => {
            const dispatcher = jest.fn()
            dispatchToProps(dispatcher).push('first value')
            expect(dispatcher).toHaveBeenCalledWith(TestItemActions.push('first value'))
        })
    })

    describe('component', () => {
        it('renders shallow', () => {
            const element = shallow(<FirstTestContent push={jest.fn()} pop={jest.fn()} selections={42} lastSelection="another value" />)
            expect(element).toMatchSnapshot()
        })
    })
})
