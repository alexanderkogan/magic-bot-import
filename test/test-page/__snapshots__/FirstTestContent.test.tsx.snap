// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstTestContent component renders shallow 1`] = `
<Component>
  <Component
    clicked={[MockFunction]}
    type="primary"
  >
    delete last item 
    <Component
      type="trash"
    />
  </Component>
  <Component />
  items selected: 
  42
  last selection: 
  another value
  <Component
    changed={[MockFunction]}
    currentValue="another value"
    id="valueToggle"
    values={
      Array [
        "first value",
        "second value",
        "another value",
      ]
    }
  />
</Component>
`;
